File: VoiceType/TranscriptionService.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace VoiceType;

/// <summary>
/// Handles OpenAI API calls: Whisper for transcription and GPT-4o for text cleanup.
/// </summary>
public class TranscriptionService : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly AppConfig _config;

    private const string WhisperEndpoint = "https://api.openai.com/v1/audio/transcriptions";
    private const string ChatEndpoint = "https://api.openai.com/v1/chat/completions";

    public TranscriptionService(AppConfig config)
    {
        _config = config;
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
        _httpClient.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Bearer", config.OpenAIApiKey);
    }

    /// <summary>
    /// Send recorded WAV audio to the Whisper API and return the raw transcription.
    /// </summary>
    public async Task<string> TranscribeAsync(byte[] audioData)
    {
        using var content = new MultipartFormDataContent();

        var audioContent = new ByteArrayContent(audioData);
        audioContent.Headers.ContentType = new MediaTypeHeaderValue("audio/wav");

        content.Add(audioContent, "file", "recording.wav");
        content.Add(new StringContent(_config.WhisperModel), "model");
        content.Add(new StringContent("en"), "language");

        var response = await _httpClient.PostAsync(WhisperEndpoint, content);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(json);
        return doc.RootElement.GetProperty("text").GetString() ?? "";
    }

    /// <summary>
    /// Send raw transcription to GPT-4o for light grammar/punctuation cleanup.
    /// </summary>
    public async Task<string> CleanupTextAsync(string rawText)
    {
        var requestBody = new
        {
            model = _config.CleanupModel,
            messages = new object[]
            {
                new { role = "system", content = _config.CleanupPrompt },
                new { role = "user", content = rawText }
            },
            max_tokens = 2048,
            temperature = 0.3
        };

        var json = JsonSerializer.Serialize(requestBody);
        var httpContent = new StringContent(json, Encoding.UTF8, "application/json");

        var response = await _httpClient.PostAsync(ChatEndpoint, httpContent);
        response.EnsureSuccessStatusCode();

        var responseJson = await response.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(responseJson);

        return doc.RootElement
            .GetProperty("choices")[0]
            .GetProperty("message")
            .GetProperty("content")
            .GetString() ?? rawText;
    }

    public void Dispose()
    {
        _httpClient.Dispose();
    }
}

File: VoiceTypeMac/Sources/TranscriptionService.swift
import Foundation

/// Handles OpenAI API calls: Whisper for transcription and GPT-4o for text cleanup.
class TranscriptionService {
    private let config: AppConfig

    private let whisperURL = URL(string: "https://api.openai.com/v1/audio/transcriptions")!
    private let chatURL    = URL(string: "https://api.openai.com/v1/chat/completions")!

    init(config: AppConfig) {
        self.config = config
    }

    // ── Whisper transcription ───────────────────────────────────────

    /// Send recorded WAV audio to the Whisper API and return the raw transcription.
    func transcribe(audioData: Data) async throws -> String {
        let boundary = UUID().uuidString

        var request = URLRequest(url: whisperURL)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("Bearer \(config.openAIApiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.appendMultipart(boundary: boundary, name: "file", filename: "recording.wav",
                             contentType: "audio/wav", data: audioData)
        body.appendMultipart(boundary: boundary, name: "model", value: config.whisperModel)
        body.appendMultipart(boundary: boundary, name: "language", value: "en")
        body.append("--\(boundary)--\r\n")

        request.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        return json?["text"] as? String ?? ""
    }

    // ── GPT-4o text cleanup ─────────────────────────────────────────

    /// Send raw transcription to GPT-4o for light grammar/punctuation cleanup.
    func cleanupText(_ text: String) async throws -> String {
        var request = URLRequest(url: chatURL)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("Bearer \(config.openAIApiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "model": config.cleanupModel,
            "messages": [
                ["role": "system", "content": config.cleanupPrompt],
                ["role": "user",   "content": text]
            ],
            "max_tokens": 2048,
            "temperature": 0.3
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let choices = json?["choices"] as? [[String: Any]]
        let message = choices?.first?["message"] as? [String: Any]
        return message?["content"] as? String ?? text
    }

    // ── Helpers ─────────────────────────────────────────────────────

    private func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200...299).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? "(no body)"
            throw VoiceTypeError.apiError(statusCode: http.statusCode, message: body)
        }
    }
}

// ── Error type ──────────────────────────────────────────────────────

enum VoiceTypeError: LocalizedError {
    case apiError(statusCode: Int, message: String)

    var errorDescription: String? {
        switch self {
        case .apiError(let code, let msg):
            return "OpenAI API error (\(code)): \(msg)"
        }
    }
}

// ── Data helpers for multipart form encoding ────────────────────────

extension Data {
    mutating func append(_ string: String) {
        if let data = string.data(using: .utf8) {
            append(data)
        }
    }

    mutating func appendMultipart(boundary: String, name: String, value: String) {
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n")
        append("\(value)\r\n")
    }

    mutating func appendMultipart(boundary: String, name: String, filename: String,
                                  contentType: String, data: Data) {
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"\(name)\"; filename=\"\(filename)\"\r\n")
        append("Content-Type: \(contentType)\r\n\r\n")
        append(data)
        append("\r\n")
    }
}
